### PR TITLE
add dead-store rule

### PR DESCRIPTION
**Pull request by inverno**
contributes to #2

### CODE REVIEW

File: tslint-sonarts.json [hunk 1]
```diff
@@ -15,6 +15,7 @@
     "no-misspelled-operator": true,
     "no-inconsistent-return": false,
     "no-unconditional-jump": true,
-    "no-misleading-array-reverse": true
+    "no-misleading-array-reverse": true,
+    "no-dead-stores": true
```
Comment (vilchik-elena) [inline tslint-sonarts.json]: @inverno I think we better use single here (to be consistent): `no-dead-store`

File: src/cfg/cfg.ts [hunk 2]
```diff
@@ -54,7 +54,7 @@ export class ControlFlowGraph {
     return this.blocks.find(block => block.loopingStatement === loopNode);
   }
 
-  private collapseEmpty() {
+  public collapseEmpty() {
```
Comment (vilchik-elena) [inline src/cfg/cfg.ts]: @inverno why?

File: src/symbols/builder.ts [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import * as tslint from "tslint";
+import * as ts from "typescript";
+import { SymbolTable, UsageFlag } from "./table";
+
+export class SymbolTableBuilder extends tslint.SyntaxWalker {
```
Comment (vilchik-elena) [inline src/symbols/builder.ts]: @inverno can't we avoid using tslint api?

File: tslint-sonarts.json (hunk 1 above)
Comment (inverno) [inline tslint-sonarts.json]: :+1: 
I had thought the same (and indeed the test files were still singular) but then something convinced me of the plural name. I don't remember what anymore, so switched back to singular as you suggest.

File: src/rules/noDeadStoresRule.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import * as tslint from "tslint";
+import * as ts from "typescript";
+import { SonarRuleMetaData } from "../sonarRule";
+//import { SymbolTableBuilder } from "../symbols/builder";
```
Comment (pynicolas) [inline src/rules/noDeadStoresRule.ts]: Commented-out code. Is there a rule to detect that?

File: src/symbols/builder.ts [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    this.addVariable(node);
+    super.visitVariableDeclaration(node);
+  }
+
+  protected visitParameterDeclaration(node: ts.ParameterDeclaration) {
+    this.addVariable(node);
+    super.visitParameterDeclaration(node);
+  }
+
+  protected visitFunctionDeclaration(node: ts.FunctionDeclaration) {
```
Comment (pynicolas) [inline src/symbols/builder.ts]: Is that useful for dead store?

File: src/symbols/builder.ts [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+      });
+    } else if (declarationName.kind === ts.SyntaxKind.ObjectBindingPattern) {
+      declarationName.elements.forEach(element => {
+        this.addVariable(element);
+      });
+    }
+  }
+
+  private registerUsage(node: ts.Node, flags: UsageFlag): ts.Symbol {
+    if (node.kind !== ts.SyntaxKind.ParenthesizedExpression) {
```
Comment (pynicolas) [inline src/symbols/builder.ts]: Style: I would change this expression to === and swap the blocks.

File: src/rules/noDeadStoreRule.ts [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    sourceFile: ts.SourceFile,
+    options: tslint.IOptions,
+    program: ts.Program,
+    private readonly symbols: SymbolTable,
+  ) {
+    super(sourceFile, options, program);
+  }
+
+  protected visitNode(node: ts.Node): void {
+    if (is(node, ...FUNCTION_LIKE)) {
```
Comment (vilchik-elena) [inline src/rules/noDeadStoreRule.ts]: @inverno I don't like the name of constant `FUNCTION_LIKE`, they are not **like** functions, they **are** functions.

File: src/symbols/builder.ts [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+export class SymbolTableBuilder extends tslint.SyntaxWalker {
+  private table = new SymbolTable();
+
+  public static build(sourceFile: ts.SourceFile, program: ts.Program): SymbolTable {
+    const builder = new SymbolTableBuilder(program);
+    builder.walk(sourceFile);
+    return builder.table;
+  }
+
+  private constructor(private program: ts.Program) {
```
Comment (vilchik-elena) [inline src/symbols/builder.ts]: @inverno should we add `readonly` for parameter?

File: src/utils/navigation.ts [hunk 9]
```diff
@@ -25,6 +25,20 @@ export function keyword(node: ts.BreakOrContinueStatement | ts.ThrowStatement |
...
 }
 
+export function isAssignment(node: ts.Node): node is ts.BinaryExpression {
+  return (
+    node.kind === ts.SyntaxKind.BinaryExpression &&
+    (node as ts.BinaryExpression).operatorToken.kind === ts.SyntaxKind.EqualsToken
+  );
+}
+
+export function retrievePureIdentifier(node: ts.Node): ts.Identifier | undefined {
```
Comment (vilchik-elena) [inline src/utils/navigation.ts]: @inverno The name of the function doesn't reflect the functionality. And I would not put such narrow function to a shared module. Could we generalise it?

File: src/cfg/cfg.ts (hunk 2 above)
Comment (inverno) [inline src/cfg/cfg.ts]: @vilchik-elena probably some forgotten change. I changed it back to private.

File: src/symbols/builder.ts (hunk 3 above)
Comment (inverno) [inline src/symbols/builder.ts]: @vilchik-elena we could. But I'm not really willing to re-implement a generic syntax walker right now (nor writing yet another specific walker). Anyway, yes, we should implement our own.

File: src/rules/noDeadStoresRule.ts (hunk 4 above)
Comment (inverno) [inline src/rules/noDeadStoresRule.ts]: @pynicolas I don't think so

File: src/symbols/builder.ts (hunk 5 above)
Comment (inverno) [inline src/symbols/builder.ts]: @pynicolas No, it's useless, as well as class, interfaces, enums and modules. I must admit that I ran amok a bit with the symbol table.

File: src/symbols/builder.ts (hunk 6 above)
Comment (inverno) [inline src/symbols/builder.ts]: @pynicolas you were probably right. Code is now completely different

File: src/utils/navigation.ts (hunk 9 above)
Comment (inverno) [inline src/utils/navigation.ts]: @vilchik-elena I've also refactored it to use `drillDownThroughParenthesis`

**Pull request merged**